### ISSUE TITLE
DEV.to header and footer displaying on Rails app with the same port

### ISSUE THREAD

**Issue by markmead** [forem/forem#10293]
**Describe the bug**

After working on Forem locally and then switching to a new Rails app on the same port, the DEV.to header and footer still show.

**To Reproduce**

1. Start Forem up
2. Close the server
3. Open a Rails app on the same server

**Expected behavior**

Not maintain cache

**Screenshots**

<img width="1425" alt="Screenshot 2020-09-12 at 11 06 11" src="https://user-images.githubusercontent.com/50486078/92993583-9f475c00-f4ea-11ea-977c-e8671e5291a8.png">

**Desktop (please complete the following information):**

- OS, version: Catalina, 10.15.6
- Browser, version: Chrome, 85

**Additional context**

- This doesn't happen on Safari
- Probably isn't limited to Rails apps on the same port

**Comment by fdoxyz** [comment]
Thank you for the report @markmead. This is a known issue and I believe this must be the service worker in place that once it's installed on `localhost:3000` it has no way of knowing whether the local development server has changed.

At the moment I'm not sure if there's a way to go around this. The only solution is to refresh without cache (on macOS `CMD + SHIFT+ R`) could be an easy way to clear out the header & footer, though I haven't tried it myself yet.

I'm leaving the issue open for now to see if anyone has an idea on how to work this out.

**Comment by markmead** [comment]
> Thank you for the report @markmead. This is a known issue and I believe this must be the service worker in place that once it's installed on `localhost:3000` it has no way of knowing whether the local development server has changed.
> 
> At the moment I'm not sure if there's a way to go around this. The only solution is to refresh without cache (on macOS `CMD + SHIFT+ R`) could be an easy way to clear out the header & footer, though I haven't tried it myself yet.
> 
> I'm leaving the issue open for now to see if anyone has an idea on how to work this out.

**Comment by markmead** [comment]
> At the moment I'm not sure if there's a way to go around this. The only solution is to refresh without cache (on macOS `CMD + SHIFT+ R`) could be an easy way to clear out the header & footer, though I haven't tried it myself yet.

On the new app, you can unregister the service worker in the dev tools - that's a more permanent fix than hard-refreshing.

**Comment by fdoxyz** [comment]
Yeah I don't think `onbeforeunload` would work because the URL would still be `localhost:3000`. That would unregister the SW anytime you're visiting external sites, unless I'm not following the idea how it could be used to unregister it.

After a quick search [I found an alternative](https://github.com/GoogleChromeLabs/sw-precache/issues/188#issuecomment-248636350) to copy&paste in the dev tools console if it feels easier:

```
navigator.serviceWorker.getRegistration().then(r => r.unregister())
```

Though I think it's more work to lookup+copy+paste in the console compared to simply unregistering the SW from the dev tools.

**Comment by markmead** [comment]
From my test, the `onbeforeunload` was working. Don't think it's the right solution though.

**Comment by jwkicklighter** [comment]
One possible solution would be to use an alias for local development rather than just `localhost`. I do this with a few projects that rely on cookies and/or JS storage so that I don't clutter up the storage with conflicting data. Each project gets an alias in my hosts file so that the cookies and such are stored under a unique domain name instead of localhost. Obviously a little more work, but there should be a way to automate changing the hosts file. It's also an easy change that you only have to do once, so a simple instruction could suffice.

**Comment by julianrubisch** [comment]
Sorry to chime in here unasked, but hijacking a contributor‘s localhost serviceworker feels borderline unacceptable.

Were we talking about a normal corporate application where this kind of setup is documented accordingly, that’d be another story. 

But unknowingly injecting code in a dev‘s browser - even more so a junior who might not immediately grok what’s going on and has to spend X hours bringing it up again - I don’t know. Asking everyone who wants to contribute to edit their /etc/hosts is likely to repulse people - it certainly would prevent me from forking and contributing. 

I think a hard configuration switch like dev:cache for local development would be the optimal solution.

**Comment by jwkicklighter** [comment]
Those are definitely valid concerns, I was suggesting the hosts file edit based on how Vagrant used to do things on projects I was part of. It was part of the startup/shutdown script of the Vagrant environment to add/remove a line to your hosts file so you had a domain name instead of localhost. While I can understand issue (and also don't know that it's the best solution, just an idea to think on), it's not unheard of and was even common practice for certain dev environments.

**Comment by julianrubisch** [comment]
Sorry, I didn’t mean to imply any bad intent. I work with service workers on a regular basis and of course you need local access for debugging. I would just advocate for not overcomplicating things for the contributing developer: make the installation/activation of the worker dependent on a config switch of some kind.

I can see how that would actually lead to more work internally, but would benefit everyone wanting to contribute value to the actual app!

**Comment by jwkicklighter** [comment]
Good clarification, that definitely makes sense.

**Comment by benhalpern** [comment]
I made this PR to give the option to skip service workers in development via `ENV` var

https://github.com/forem/forem/pull/10617

**Comment by DaveParr** [comment]
FWIW I think I'm currently experiencing this issue as I've been teaching myself react. It's really confused me (I've not a lot of front end experience yet):

What I did:
Set up the Dev.to environment when you went open source a few months ago, then started a new job and didn't get much further
Went back to the same laptop and started doing React tutorials
```sh
npx create-react-app my-app
cd my-app
npm start
```
Open localhost:3000 
What I expected to see:
<img width="1280" alt="Screenshot 2021-01-30 at 18 15 14" src="https://user-images.githubusercontent.com/8363743/106364478-29bc1b00-6327-11eb-9947-81d9533fd5e3.png">

What I actually saw:
<img width="1280" alt="Screenshot 2021-01-30 at 18 16 11" src="https://user-images.githubusercontent.com/8363743/106364486-435d6280-6327-11eb-9643-3b95306250f3.png">

Though I've repeatedly used the cmd + shift + r to clear the cache when the localhost is started, the issue occurs _everytime_ I kill the terminal process and start over, and in multiple projects. I've done 3 tutorials now and this issue has turned up in every single one.

I've now run:
```sh
navigator.serviceWorker.getRegistration().then(r => r.unregister())
```
as described above,  which seems to have worked, though I don't really understand yet what it did or why this is a problem.

TL;DR: +1 on not doing this in the future.

That said, thanks as always for making a great community. Really value what dev.to does, and was really excited to get some dev.to stickers for hacktoberfest this year 👍

**Comment by jacobherrington** [comment]
Hey @DaveParr, there really isn't much of a way to avoid this at the moment. Of course, we should consider making some changes down the road to make this less painful.

For the time being, if you want to learn more about Service Workers (I'm still learning about them myself), [Google has a pretty good introduction to them](https://developers.google.com/web/fundamentals/primers/service-workers) or you can check out this brief [DEV article](https://dev.to/attacomsian/introduction-to-server-workers-in-javascript-1epa).

I think Ben's PR to disable service workers by default in the development environment should make this a less common occurrence until we come up with a good alternative.

**Comment by citizen428** [comment]
@DaveParr If you have some backend experience, the simplest explanation I can give for a service worker is "like a background job running in the browser".

BTW: You can also unregister service workers from the developer tools. Go to the "Application" tab (where you can see things like service workers, local storage, etc), select "Service Workers" in the left sidebar, and hit "Unregister": 

![image](https://user-images.githubusercontent.com/47985/106544330-ea194d00-6539-11eb-9a04-932ca7bd29f3.png)

This achieves the same as `navigator.serviceWorker.getRegistration().then(r => r.unregister())` but is a bit less cryptic and you may find other neat things in this tab, especially if you're just getting started with frontend development.

**Comment by mstruve** [comment]
Since the service worker has been removed this should no longer be a problem! Please feel free to reopen if it still is. Thanks!
